open-liberty/content.md

```diff
@@ -0,0 +1,148 @@
+
+1.	Each image contains a default server configuration that specifies the corresponding features and exposes ports 9080 and 9443 for HTTP and HTTPS respectively. A .WAR file can therefore be mounted in the `dropins` directory of this server and run. The following example starts a container in the background running a .WAR file from the host file system with the HTTP and HTTPS ports mapped to 80 and 443 respectively.
+
+	```console
+	$ docker run -d -p 80:9080 -p 443:9443 \
+	    -v /tmp/DefaultServletEngine/dropins/Sample1.war:/config/dropins/Sample1.war \
+	    %%IMAGE%%:webProfile7
+	```
+
+	When the server is started, you can browse to http://localhost/Sample1/SimpleServlet on the Docker host.
+
+2.	For greater flexibility over configuration, it is possible to mount an entire server configuration directory from the host and then specify the server name as a parameter to the run command. Note: This particular example server configuration provides only HTTP access.
+
+	```console
+	$ docker run -d -p 80:9080 \
+	  -v /tmp/DefaultServletEngine:/config \
+	  %%IMAGE%%:webProfile7
+	```
+
+3.	You can also build an application layer on top of this image by using either the default server configuration or a new server configuration. In this example, we have copied the `Sample1.war` from `/tmp/DefaultServletEngine/dropins` to the same directory as the following Dockerfile.
+
+	```dockerfile
+	FROM %%IMAGE%%:webProfile7
+	ADD Sample1.war /config/dropins/
+	```
+
+	This can then be built and run as follows:
+
+	```console
+	$ docker build -t app .
+	$ docker run -d -p 80:9080 -p 443:9443 app
+	```
+
+4.	You can mount a data volume container that contains the application and the server configuration on to the image. This has the benefit that it has no dependency on files from the host but still allows the application container to be easily re-mounted on a newer version of the application server image. This example assumes that you have copied the `/tmp/DefaultServletEngine` directory in to the same directory as the Dockerfile.
+
+	Build and run the data volume container:
+
+	```dockerfile
+	FROM %%IMAGE%%:webProfile7
+	ADD DefaultServletEngine /config
+	```
+
+	```console
+	$ docker build -t app-image .
+	$ docker run -d -v /config \
+	    --name app app-image true
+	```
+
+	Run the Open Liberty image with the volumes from the data volume container mounted:
+
+	```console
+	$ docker run -d -p 80:9080 \
+	  --volumes-from app %%IMAGE%%:webProfile7
+	```
+
+# Providing your own keystore/truststore
+
+When an `open-liberty` image starts, it can generate a Liberty server XML snippet in `/config/configDropins/defaults/keystore.xml` that specifies a `keyStore` stanza with a generated password. This causes Open Liberty to generate a default keystore and truststore with a self-signed certificate when it starts. The `javaee7` image does this automatically, but other images can request this by setting:
+
+```console
+ENV KEYSTORE_REQUIRED "true"
```

Comment:
This particular variable name seems to be unique to open-liberty -- any particular reason for it?  In websphere-liberty, this is just done unilaterally, which should be fine, right?

Code from https://github.com/WASdev/ci.docker/blob/9dc5c2e942d20ddc2f913ae0a0dbcc9d174afc00/ga/developer/kernel/docker-server#L22-L33:

```bash
keystorePath="/config/configDropins/defaults/keystore.xml"

if [ ! -e $keystorePath ]
then
  # Generate the keystore.xml
  PASSWORD=$(openssl rand -base64 32)
  XML="<server description=\"Default Server\"><keyStore id=\"defaultKeyStore\" password=\"$PASSWORD\" /></server>"

  # Create the keystore.xml file and place in configDropins
  mkdir -p $(dirname $keystorePath)
  echo $XML > $keystorePath
fi
```

Reply:
This is unique to Open Liberty image. There are a few issues with what the websphere-liberty does. The main one is that most people who use the websphere-liberty image create their own Dockerfile with a FROM clause. In our experience people who care about security tend to have their own certificates from some kind of CA. When they provide those in their Dockerfile they will have their own certificates. That means that the step the websphere-liberty driver has to be 'disabled' which requires them to put their SSL configuration in /config/configDropins/defaults/keystore.xml which isn't exactly intuitive. We have tried very hard not to force conventions on our users when naming their configuration files.

Another advantage this has is the creation of this password causes the app server to generate an SSL certificate during startup. We do this because it makes it easier to do SSL out of the box than requiring someone to create a self signed cert and knowing what config to do for a JVM, however it also affects server startup time, so docker run takes longer. If you don't require SSL then that is an overhead you don't want. The intent of making open-liberty differ from websphere-liberty was to allow an environment variable to dictate whether this happens or not. This also feels to me to be more in keeping with the 3rd factor for 12 factor apps.